Add ImageRecordReader constructors taking PathLabelGenerator, deprecating others

saudet opened the pull request:
Fixes #132? For #135...

nyghtowl commented:
- Appears BaseImageRR and ImageRR can be collapsed
- Constructor removed that allows passing in list of labels which is valuable when testing a small subset of the data.
- Appendlabel is set to true but for unsupervised no labels are needed. Arguably a constructor could be set to say if a PathGenerator is left out or is null (same with list of labels if that is left) then appendlabel should be false.

- In deprecating these constructors can you go through ComputerVisionExamples and share how they would look with the new design (esp. [MRSA](https://github.com/deeplearning4j/ComputerVision-examples/blob/master/src/main/java/org/deeplearning4j/examples/cv/msra_cfw/MSRA_CFW.java#L113))

saudet commented:
With MSRA it would look like this:
```java
RecordReader recordReader = new ImageRecordReader(HEIGHT, WIDTH, CHANNELS, new ParentPathLabelGenerator());
```
But what I'm wondering most about is if there are times when we need to provide the whole set of available labels in advance and under what circumstances would that be necessary?

saudet commented:
After discussion with @nyghtowl I've tweaked a few things. When specifying maxPathsPerLabel, BalancedPathFilter now tries to return that much examples per label, even if there are fewer than that for some. Thus might not be fully balanced, but more useful that way... I've also added a "labels" argument to specify manually the labels to accept, basically taking on a potential role for the list that was traditionally passed to the ImageRecordReader directly. That basically takes care of my concern as well. So, good to merge?

And about collapsing BaseImageRecordReader into ImageRecordReader, I think it's alright the way it is now, but obviously we should eventually move less general stuff to ImageRecordReader...

nyghtowl commented on canova-api/src/main/java/org/canova/api/io/filters/BalancedPathFilter.java:
```diff
@@ -87,14 +107,16 @@ public BalancedPathFilter(Random random, String[] extensions, PathLabelGenerator
             }
         }
-        if (maxPathsPerLabel > 0 && minCount > maxPathsPerLabel) {
+        if (maxPathsPerLabel > 0) {
```
Ideal to keep both - an option just to balance based on maxPath so there can be different sizes up to a number and an option that will require all classes to have the exact same number.

saudet commented:
Ok, I've updated BalancedPathFilter to restore old behavior with maxPathsPerLabel, and added a new minPathsPerLabel one to control the lower end, such that if we set minPathsPerLabel = maxPathsPerLabel, it'll try to return that amount, might return less, but won't return more.
